wsmaster/che-core-api-workspace-activity/src/main/java/org/eclipse/che/api/workspace/activity/WorkspaceActivityManager.java

```diff
@@ -100,44 +106,44 @@ public void onEvent(WorkspaceStatusEvent event) {
    * @param activityTime moment in which the activity occurred
    */
   public void update(String wsId, long activityTime) {
-    try {
-      long timeout = getIdleTimeout(wsId);
-      if (timeout > 0) {
-        activeWorkspaces.put(wsId, activityTime + timeout);
-      }
-    } catch (NotFoundException | ServerException e) {
-      LOG.error(e.getLocalizedMessage(), e);
-    }
-  }
-
-  protected long getIdleTimeout(String workspaceId) throws NotFoundException, ServerException {
     if (timeout > 0) {
-      return timeout;
-    } else {
-      return -1;
+      try {
+        activityDao.setExpiration(new WorkspaceExpiration(wsId, activityTime + timeout));
+      } catch (ServerException e) {
+        LOG.error(e.getLocalizedMessage(), e);
+      }
     }
   }
 
-  @ScheduleRate(periodParameterName = "che.workspace.activity_check_scheduler_period_s")
+  @ScheduleDelay(
+    initialDelayParameterName = "che.workspace.activity_check_scheduler_delay_s",
+    delayParameterName = "che.workspace.activity_check_scheduler_period_s"
+  )
   private void invalidate() {
-    final long currentTime = System.currentTimeMillis();
-    for (Map.Entry<String, Long> workspaceExpireEntry : activeWorkspaces.entrySet()) {
-      if (workspaceExpireEntry.getValue() <= currentTime) {
+    List<String> expiredList = new ArrayList<>();
+    try {
+      expiredList = activityDao.findExpired(System.currentTimeMillis());
+    } catch (ServerException e) {
+      LOG.error(e.getLocalizedMessage(), e);
```

Comment:
In this case, the flow is kinda interrupted since `expiredList` will be empty. Maybe it makes sense to put return here to explicitly show this?

Reply:
Makes sense

Reply:
Why don't you combine this try with try below?
```
try{
 activityDao.findExpired(System.currentTimeMillis()).forEach(this::stopExpired);
}
```